### PR TITLE
Ports - Amy M

### PR DESCRIPTION
**Pull request by amyesh**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | Finding the top ten media for each category.
Describe how you approached testing that model method. What edge cases did you come up with? | For nominal tests I tested that the resulting array was sorted in descending order by counting the votes of the first and last work, and that it returned only 10 works, and for edge cases I tested that it will return an empty array if there are no works, and that in the case of no votes, it will return a list of the first ten works.
What are `session` and `flash`? What is the difference between them? | Session and flash are both hashes, where session stores data about the users browser session, flash stores data for one-time use to send a message from controller to the view, and unlike session will go away after the next request-response cycle
What was one thing that you gained more clarity on through this assignment? | the rails architecture/magic in general.
What is the Heroku URL of your deployed application? | https://medium-ranker.herokuapp.com/ https://trello.com/b/NmiyjvvI/mediaranker
Do you have any recommendations on how we could improve this project for the next cohort? |  I think fixtures added a layer of confusion and complexity to testing that made this project a lot harder to get through. I'm still unclear about exactly why they need to be called into a variable name when they always exist? I feel like I might've just needed more time with learning about how let and before/do tie in with fixtures before jumping into testing in rails. The combination of rails magic and fixture magic was...extra magical?!

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | x
Comprehension questions | x
**General** |
Rails fundamentals (RESTful routing, use of named paths) | x
Views are well-organized (DRY, use of semantic HTML, use of partials) | x, opportunity to refactor the homepage and the works index view
Errors are reported to the user | x
Business logic lives in the models | x
Models are thoroughly tested, including relations, validations and any custom logic | x
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | x
**Wave 1 - Media** |
Splash page shows the three media categories | x
Basic CRUD operations on media are present and functional | x
**Wave 2 - Users and Votes** |
Users can log in and log out | x
The ID of the current user is stored in the session | x
A user cannot vote for the same media more than once | x
All media lists are ordered by vote count | x
Splash page contains a media spotlight | x
**Wave 3 - Users and Votes** | 
Media pages contain lists of voting users | x
Individual user pages and the user list are present | x
**Optional - Styling** |
Bootstrap is used appropriately | x
Look and feel is similar to the original | x
**Overall** |



Great work on this, Amy! Your project looks good: you got all of the requirements and functionality, and this was a huge project. Great work particularly on code style and the tests.


I have really only a few comments on how some code could be refactored.


Otherwise, overall, great work on this project. Well done!

File: app/models/work.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  validates :category, presence: true, inclusion: {in: %w(album movie book)}
+
+  def self.top_media
+    if self.count == 0
+      top_media = nil
+    else
+      works = self.all.sort_by { |work| work.votes.length }.reverse!
+      top_media = works.first
+    end
+    return top_media
```
Comment (tildeee) [inline app/models/work.rb:14]: _Do_ we need to check if `self.count` is zero?

Consider this: calling `.first` on an empty array returns `nil` (aka in IRB, we can check: `[].first` returns `nil`)

So consider this refactor, which passes all of the tests still:

```ruby
  def self.top_media
    works = self.all.sort_by { |work| work.votes.length }.reverse!
    top_media = works.first
    return top_media
  end
```

File: test/controllers/votes_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
+require "test_helper"
+
+describe VotesController do
+  describe "create" do
+    before do
+      @user = perform_login
```
Comment (tildeee) [inline test/controllers/votes_controller_test.rb:6]: Nice setup!

File: test/controllers/votes_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      @work = works(:two)
```
Comment (tildeee) [inline test/controllers/votes_controller_test.rb:7]: In response to your question on the comprehension questions: you don't _need_ to assign `@work = works(:two)` ... it's just an option, so that you don't have to do `works(:two)` each time (because `@work` is shorter?? I guess?? Also it vaguely reads more easily, so you don't have to think about syntax)

Comment (tildeee) [reply test/controllers/votes_controller_test.rb:7]: But yeah, you're free to just do `works(:two)` the entire time

File: test/controllers/votes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+      expect {
+        post work_votes_path(@work.id)
+      }.wont_change "Vote.count"
+
+      expect(flash[:error]).must_equal "Error - upvote unsuccessful. You can't vote for the same media twice."
+    end
+
+    it "should not allow a user to vote if they aren't logged in" do
+      post logout_path
```
Comment (tildeee) [inline test/controllers/votes_controller_test.rb:32]: yayyy this is nice and cool

File: app/views/homepages/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,46 @@
+ <section class="spotlight">
+    <% top_media = Work.top_media %>
```
Comment (tildeee) [inline app/views/homepages/index.html.erb:2]: Hm, instead of setting this local variable here, consider setting it in the `HomepagesController#index` action as an instance variable
